### PR TITLE
Scalable teardown lock for FreeBSD

### PR DESCRIPTION
**Pull request by mjguzik**
### Motivation and Context

"reader-mostly lock" (module/zfs/rrwlock.c) suffers significant scalability problems and is used all the time with ZFS_ENTER/ZFS_EXIT. The lock array of 17 elements hashed into based on current thread address very early runs into conflicts, inhibiting performance way below the 17 limit. thread address hashing has additional problem of bouncing locks across different NUMA nodes if they happen to match. All of this is fixable with enough effort, but I have no immediate interest in reimplementing it.

Effect can be seen here: https://people.freebsd.org/~mjg/fstat1-zfs.svg

This patchset switches the teardown lock to use a scalable routine present in the FreeBSD kernel.

Non-FreeBSD changes are almost a no-op and mostly concern abstracting teardown_lock use with macros. Note with the patchset in place it will also be possible to fix the problem on Linux with whatever Linux-specific manner available.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by freqlabs — APPROVED** [review]

**Comment by mjguzik** [comment]
Turns out there is a funny FreeBSD bug here (assertion-related), I'll be updating the review some time over the weekend.

**Review by mattmacy — APPROVED** [review]

File: include/os/freebsd/zfs/sys/zfs_vfsops_os.h [hunk 1]
```diff
@@ -46,10 +50,16 @@
...
 typedef struct rmslock zfs_teardown_lock_t;
 #else
-#define	zfs_teardown_lock_t krwlock_t
+#define	zfs_teardown_lock_t		rrmlock_t
+#endif
+
+#ifdef TEARDOWN_INACTIVE_RMS
+typedef struct rmslock zfs_teardown_inactive_lock_t;
+#else
+#define	zfs_teardown_inactive_lock_t krwlock_t
```
Comment (behlendorf) [inline include/os/freebsd/zfs/sys/zfs_vfsops_os.h:62]: One question, could we instead tackle this issue by mapping the existing `rrw_*` / `rrm_*` interfaces to the right FreeBSD locking primitive in the same way the other locks types are handled (mutexes, rwlocks)?  Or are the semantics and interfaces just too different to accommodate this.

**Review by behlendorf — COMMENTED** [review]
> Note with the patchset in place it will also be possible to fix the problem on Linux with whatever Linux-specific manner available.

That raises a good point.  The ideal way to address this for Linux would be to remove the lock entirely.  Since each mounted filesystem holds a reference on the super block, and each process holds a reference on the filesystem namespace, there's really not a need for the lock and the reference count could be used.  That said, getting there I'm sure would take a fair bit more work so it's definitely out of scope for this.  Incidentally, it's this design decision to use a reference instead of a lock which prevents a force unmount on Linux.

**Comment by mjguzik** [comment]
FreeBSD snapshots failed to generate last week and are needed to test the patchset. Pending change to this PR just removes 2 low quality asserts which are not reliable with the new primitive. I'm going to push it after the snapshot is generated, which should be in 2 days.

> > Note with the patchset in place it will also be possible to fix the problem on Linux with whatever Linux-specific manner available.
> 
> That raises a good point. The ideal way to address this for Linux would be to remove the lock entirely. Since each mounted filesystem holds a reference on the super block, and each process holds a reference on the filesystem namespace, there's really not a need for the lock and the reference count could be used. That said, getting there I'm sure would take a fair bit more work so it's definitely out of scope for this. Incidentally, it's this design decision to use a reference instead of a lock which prevents a force unmount on Linux.

Interestingly the code is almost mostly useless on FreeBSD as the vnode locking protocol from the VFS layer keeps forced unmount from progressing until you return from fs ops. However, there is administrative stuff which takes the lock and it may be that matters.

File: include/os/freebsd/zfs/sys/zfs_vfsops_os.h (hunk 1 above)
Comment (mjguzik) [reply include/os/freebsd/zfs/sys/zfs_vfsops_os.h:62]: rrm feature-wise it supports tracking who has the read lock so you can assert on it if you want and allows certain read-locks attempts to progress despite pending write waiters, neither of which are implemented in rms (it can be done but I have not seen any need to do it). Apart from that rrm, despite its performance woes, offers a different tradeoff for write locking -- take up to n locks and be done with it, while rms sends an IPI to all CPUs. Tolerable for the time being for unmount or other relatively rare events, but may not be ok for whatever upcoming rms use may show up.

tl;dr too different :)

**Comment by mjguzik** [comment]
FreeBSD head test above failed somewhere in tests/functional/rsend. I ran it locally and it passes no problem, thus I suspect a transient issue. However, if there is a way to get dmesg that would help.

**Comment by freqlabs** [comment]
Unfortunately we don't have a way to get dmesg for failures like that where typically either a deadlock or a panic occurred.

**Comment by mjguzik** [comment]
So what are the next steps? Can the build bot be tasked with building this again? Do I need to do anything here?

**Comment by behlendorf** [comment]
@mjguzik I've resubmitted the FreeBSD HEAD failure to see if it's reproducible in the CI.

**Comment by mmatuska** [comment]
Any news on this getting in? We have this in FreeBSD-HEAD since November and now it is going to be released in 13.0-RELEASE.

**Comment by mmatuska** [comment]
@behlendorf any objections to this?

**Comment by behlendorf** [comment]
@mmatuska @mjguzik sorry, I lost track of this one.  My only concern is that it slightly conflicts with the change #11082.  However, it looks like it'll be straight forward to update that change to use the new macros so it shouldn't be a problem.  Can we just get this rebased on the current master source to get a new test run before merging it.

**Review by behlendorf — APPROVED** [review]
